### PR TITLE
344 - Lookup API and E2E Tests

### LINKED ISSUE #344: Lookup Functional and E2E Tests
Add some tests for Lookup

See existing tests for example textarea, dropdown, treemap, menubutton for reference. Also see docs/TESTING.MD for commands. Ask for help on sohoxi-dev if needed.

Functional Tests
- Test events that get triggered
- Test events that are listened to
- Test all API methods, most importantly the public ones first and the rest for coverage. (fx enable/disable)
- Test lifecycle methods (init, teardown, destroy, invoke)
- Test setting and that some states are applied

E2e Tests
- Load each test page in the components/<name>/list and test the main ideas of the page, if unsure of the main idea ask on the channel.
- Keyboard

Lookup Specific Test Considerations 
- custom buttons on the lookup
- changing toolbar
- editing in the field
- editing in the lookup
- rendering a filter row
- single vs multiselect
- autoWidth option
- events (fx before open)
- block opening
- custom click function
- matching
- changing delimiter
- auto apply
- states
- validation on the buttons
- on a cap
- on a modal
- server side paging vs client side paging
- setting selected values

Other Tests
- Accessibility tests https://github.com/dequelabs/axe-core
- Security Tests XSS
- Other Specific Test Cases from Cindy
- Protractor Image Comparison Tests

### PR DESCRIPTION
**Pull request by tjamesallen15**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**

Adding API and E2E Test for Lookup Component

**Related github/jira issue (required)**:

Closes #344

**Steps necessary to review your pull request (required)**:

npm run functional:ci
npm run e2e:ci

### CODE REVIEW

File: test/components/lookup/lookup.e2e-spec.js [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+requireHelper('rejection');
+
+jasmine.getEnv().addReporter(browserStackErrorReporter);
+
+describe('Lookup example-custom-buttons tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/lookup/example-custom-buttons');
+  });
+
+  it('Should lookup enabled', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should be enabled?

File: test/components/lookup/lookup.e2e-spec.js [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    const lookupEl = await element(by.id('custom'));
+
+    expect(await lookupEl.isEnabled()).toBe(true);
+  });
+
+  it('Should lookup display', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should display modal

File: test/components/lookup/lookup.e2e-spec.js [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    const buttonEl = await element(by.className('trigger'));
+    await buttonEl.click();
+
+    expect(await element(by.className('modal-content'))).toBeTruthy();
+  });
+
+  it('Should lookup has custom button', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should be able to have a custom button

File: test/components/lookup/lookup.e2e-spec.js [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+    const buttonEl = await element.all(by.className('trigger')).first();
+    await buttonEl.click();
+
+    expect(await element(by.className('btn-modal-primary')).getText()).toEqual('Apply It Now');
+  });
+
+  it('Should lookup set a value', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should be able to set a value

File: test/components/lookup/lookup.e2e-spec.js [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+    await buttonEl.click();
+
+    await browser.driver.wait(protractor.ExpectedConditions.presenceOf(element(by.className('datagrid-cell-wrapper'))), config.waitsFor);
+    await element.all(by.className('datagrid-cell-wrapper')).first().click();
+
+    expect(await lookupEl.getAttribute('value')).toEqual('2142201');
+  });
+
+  it('Should lookup apply button enabled', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should have the apply button enabled

File: test/components/lookup/lookup.e2e-spec.js [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+    expect(await element(by.className('btn-modal-primary')).isEnabled()).toBe(true);
+  });
+});
+
+describe('Lookup example-custom-toolbar tests', () => {
+  beforeEach(async () => {
+    await utils.setPage('/components/lookup/example-custom-toolbar');
+  });
+
+  it('Should lookup field enabled', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should be enabled?

File: test/components/lookup/lookup.e2e-spec.js [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+    const lookupEl = await element(by.id('custom'));
+
+    expect(await lookupEl.isEnabled()).toBe(true);
+  });
+
+  it('Should lookup with custom toolbar display', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should display a custom toolbar

File: test/components/lookup/lookup.e2e-spec.js [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+    const buttonEl = await element(by.className('trigger'));
+    await buttonEl.click();
+
+    expect(await element(by.className('modal-content'))).toBeTruthy();
+  });
+
+  it('Should lookup has custom toolbar', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should have a custom toolbar with a more button

File: test/components/lookup/lookup.e2e-spec.js [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+    const buttonEl = await element.all(by.className('trigger')).first();
+    await buttonEl.click();
+
+    expect(await element(by.className('toolbar'))).toBeTruthy();
+    expect(await element(by.className('has-more-button'))).toBeTruthy();
+  });
+
+  it('Should lookup with custom toolbar set a value', async () => {
```
Comment (tmcconechy) [inline test/components/lookup/lookup.e2e-spec.js]: Should be able to select a value when using a custom toolbar

**Comment by EdwardCoyle** [comment]
Agree with Tim here.  Generally we should be able to look at the results output of the test suite and figure out what functionality isn't quite working if a test fails by a plainly-spoken English description of what the functionality is.

Take a look at some of the other test suites, or this article's test defs if you're looking for some ideas:  https://github.com/CarmenPopoviciu/protractor-styleguide#e2e-testing

**Comment by tjamesallen15** [comment]
Hi @EdwardCoyle  and @tmcconechy,

The reason why some of the description are too much descriptive because eslint won't allow me to make same descriptions.

Thanks!

File: test/components/lookup/lookup.e2e-spec.js [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+    const buttonEl = await element.all(by.className('trigger')).first();
+    await buttonEl.click();
+
+    await browser.driver.wait(protractor.ExpectedConditions.presenceOf(element(by.className('modal-content'))), config.waitsFor);
+
+    expect(await element(by.className('modal-content'))).toBeTruthy();
+  });
+
+  it('Should have an enabled modal cancel button', async () => {
+    const buttonEl = await element(by.className('trigger'));
```
Comment (d-g-h) [inline test/components/lookup/lookup.e2e-spec.js]:
```sh
[16:23:06] W/element - more than one element found for locator By(css selector, .trigger) - the first result will be used
    ✓ Should have an enabled modal cancel button

```

Comment (d-g-h) [reply test/components/lookup/lookup.e2e-spec.js]: @tjamesallen15 Can we add a `.first()` here please to remove the warnings?

**Pull request merged**